node/core/prospective-parachains/src/tests.rs

```diff
@@ -1320,3 +1343,100 @@ fn correctly_updates_leaves() {
 	assert_eq!(view.active_leaves.len(), 0);
 	assert_eq!(view.candidate_storage.len(), 0);
 }
+
+#[test]
+fn persists_pending_availability_candidate() {
+	let mut test_state = TestState::default();
+	let para_id = ParaId::from(1);
+	test_state.availability_cores = test_state
+		.availability_cores
+		.into_iter()
+		.filter(|core| core.para_id().map_or(false, |id| id == para_id))
+		.collect();
+	assert_eq!(test_state.availability_cores.len(), 1);
+
+	test_harness(|mut virtual_overseer| async move {
+		let para_head = HeadData(vec![1, 2, 3]);
+
+		// Min allowed relay parent for leaf `a` which goes out of scope in the test.
+		let candidate_relay_parent = Hash::from_low_u64_be(5);
+		let candidate_relay_parent_number = 97;
+
+		let leaf_a = TestLeaf {
+			number: candidate_relay_parent_number + ALLOWED_ANCESTRY_LEN,
+			hash: Hash::from_low_u64_be(2),
+			para_data: vec![(
+				para_id,
+				PerParaData::new(candidate_relay_parent_number, para_head.clone()),
+			)],
+		};
+
+		let leaf_b_hash = Hash::from_low_u64_be(1);
+		let leaf_b_number = leaf_a.number + 1;
+
+		// Activate leaves.
+		activate_leaf(&mut virtual_overseer, &leaf_a, &test_state).await;
+
+		// Candidate A
+		let (candidate_a, pvd_a) = make_candidate(
+			candidate_relay_parent,
+			candidate_relay_parent_number,
+			para_id,
+			para_head.clone(),
+			HeadData(vec![1]),
+			test_state.validation_code_hash,
+		);
+		let candidate_hash_a = candidate_a.hash();
+
+		// Candidate B, built on top of the candidate which is out of scope but pending availability.
+		let (candidate_b, pvd_b) = make_candidate(
+			leaf_b_hash,
+			leaf_b_number,
+			para_id,
+			HeadData(vec![1]),
+			HeadData(vec![2]),
+			test_state.validation_code_hash,
+		);
+		let candidate_hash_b = candidate_b.hash();
+
+		introduce_candidate(&mut virtual_overseer, candidate_a.clone(), pvd_a).await;
+		second_candidate(&mut virtual_overseer, candidate_a.clone()).await;
+		back_candidate(&mut virtual_overseer, &candidate_a, candidate_hash_a).await;
+
+		let candidate_a_pending_av = CandidatePendingAvailability {
+			candidate_hash: candidate_hash_a,
+			descriptor: candidate_a.descriptor.clone(),
+			commitments: candidate_a.commitments.clone(),
+			relay_parent_number: candidate_relay_parent_number,
+			max_pov_size: MAX_POV_SIZE,
+		};
+		let leaf_b = TestLeaf {
+			number: leaf_b_number,
+			hash: leaf_b_hash,
+			para_data: vec![(
+				1.into(),
+				PerParaData::new_with_pending(
+					candidate_relay_parent_number + 1,
+					para_head.clone(),
+					vec![candidate_a_pending_av],
+				),
+			)],
+		};
+		activate_leaf(&mut virtual_overseer, &leaf_b, &test_state).await;
+
+		introduce_candidate(&mut virtual_overseer, candidate_b.clone(), pvd_b).await;
+		second_candidate(&mut virtual_overseer, candidate_b.clone()).await;
+		back_candidate(&mut virtual_overseer, &candidate_b, candidate_hash_b).await;
+
+		get_backable_candidate(
```

Comment:
Ah, I remember being confused by this terminology when I was first working with this file. It is still a bit weird actually, reading the test, seeing that `back_candidate` leads to the candidate being "backable". I'm not sure if you saw it @rphmeier but a couple of other newer folks also expressed that this terminology was confusing. Not necessarily saying this should be changed, but maybe something to keep in mind.

Reply:
we ought to do a separate pass for that. I agree it's confusing and often used interchangeably. @slumber added this code, actually.

Reply:
I just added a test for pending availability logic 😢